4882 invite users to your organization

juanignaciosl opened the pull request:
This closes #4882.

Testing:
- Invite a user outside whitelisted domains and check he can sign up from the email.

Regressions:
- Check organization user creation.
- Check normal organization signup.
- Check normal signup.

juanignaciosl commented:
@Kartones CR this, please. It's deployed at ded03 if you want to have a try.

juanignaciosl commented:
@xavijam feel free to use ded03 for testing frontend. I've already tested backend and I'm just waiting for your fix and for CR to deploy.

Kartones commented on app/models/carto/invitation.rb:
```diff
@@ -0,0 +1,90 @@
+
+    def use(email, token)
+      # reload and used_emails assignment is needed because otherwise
+      # activerecord-postgresql-array won't update the invitations
```
I'm not sure I would have went this way instead of just serializing to string as currently the arrays don't work fine, but good to know there is a way to force it to work.

juanignaciosl replied:
Yep, both approaches have their benefits and drawbacks. I preferred this because I can group the workaround in the factory method and then forget everywhere else, taking advantage of native arrays.

Kartones commented on lib/assets/javascripts/cartodb/models/organization.js:
```diff
@@ -81,7 +81,7 @@ cdb.admin.Organization.Users = Backbone.Collection.extend({
     exclude = !!exclude;
     this._excludeCurrentUser = exclude;
-    if (exclude && !this.currentUserId) {
+    if (exclude && !this.currentUserId) {s
```
that `s` looks like a typo, careful

juanignaciosl replied:
Good catch! Fixed. cc @xavijam ^

Kartones commented on lib/assets/javascripts/cartodb/old_common/utils.js:
```diff
@@ -254,6 +254,11 @@
     return str.charAt(0).toUpperCase() + str.slice(1);
   };
 
+  cdb.Utils.isValidEmail = function(str) {
```
I was going to tell you the same for the BE invitation model... why not a simple email validation? for those "never trust the client" security practices, for lowercasing and trimming the emails, etc.

juanignaciosl replied:
Ok, added, I'll ping you when it's pushed.

Kartones commented on app/models/carto/invitation.rb:
```diff
@@ -0,0 +1,90 @@
+      # reload and used_emails assignment is needed because otherwise
+      # activerecord-postgresql-array won't update the invitations
+      reload
+      if users_emails.include?(email) && self.token(email) == token
```
So we cannot repeat emails on the same invitation, but we can on consecutive ones, no?
I guess this is good for the case of an invitation going to the spam folder, owner can always  send another one

juanignaciosl replied:
Yep. And since invitations are one use only, you could invite a user, then delete his account, then create it afterwards. Maybe a little twisted, but I can't see a reason to block duplicated invitations.

Kartones commented:
looks good to me :+1:

juanignaciosl commented:
Retest this, please.

juanignaciosl commented:
Retest this, please.

juanignaciosl commented:
@josecruz FYI this will be released tomorrow, you can test in ded03

juanignaciosl commented:
@Kartones I added server side email format check (and fixed the typo), take a look if you want.

@xavijam @josecruz I'll merge & deploy this tomorrow morning.

Kartones commented:
:+1:

josecruz commented:
Found a small frontend bug:
![screen shot 2015-10-05 at 17 15 57](https://cloud.githubusercontent.com/assets/596509/10284848/58537ba0-6b86-11e5-93bb-2a897cd3be57.png)

Also, I'm not sure if we should show an _invite success modal_ to let the user know that the invite has been sent correctly

juanignaciosl commented:
I'm afraid that's a known bug of that widget. cc @xavijam .

saleiva commented:
We should have an state for that @josecruz something elegant

juanignaciosl commented:
@saleiva Ok. If you agree we'll deploy this tomorrow, we can apply @josecruz improved design later.

saleiva commented:
+1
